File: js/nebulachain/shared/contracts/contracts.js
'use strict';

// ---------------------------------------------------------------------------
// Service Contracts — defines the inter-service communication contracts
// for the NebulaChain dispatch governance platform.
// ---------------------------------------------------------------------------

const SERVICE_DEFINITIONS = Object.freeze({
  gateway: {
    id: 'gateway',
    port: 8090,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: ['routing', 'policy'],
  },
  routing: {
    id: 'routing',
    port: 8091,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: ['policy'],
  },
  policy: {
    id: 'policy',
    port: 8092,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: [],
  },
  resilience: {
    id: 'resilience',
    port: 8093,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: ['policy'],
  },
  analytics: {
    id: 'analytics',
    port: 8094,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: ['routing'],
  },
  audit: {
    id: 'audit',
    port: 8095,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: [],
  },
  notifications: {
    id: 'notifications',
    port: 8096,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: ['policy'],
  },
  security: {
    id: 'security',
    port: 8097,
    healthPath: '/health',
    version: '1.0.0',
    dependencies: [],
  },
});


const REQUIRED_COMMAND_FIELDS = ['commandId', 'serviceId', 'action', 'payload']; 


function getServiceUrl(serviceId, baseDomain) {
  const svc = SERVICE_DEFINITIONS[serviceId];
  if (!svc) return null;
  const domain = baseDomain || 'localhost';
  return `http://${domain}:${svc.port}`; 
}

function validateContract(serviceId) {
  const svc = SERVICE_DEFINITIONS[serviceId];
  if (!svc) return { valid: false, reason: 'unknown_service' };
  if (!svc.port || svc.port < 1024) return { valid: false, reason: 'invalid_port' };
  // Check dependencies exist
  for (const dep of svc.dependencies) {
    if (!SERVICE_DEFINITIONS[dep]) return { valid: false, reason: `missing_dependency_${dep}` };
  }
  return { valid: true, service: svc };
}


function topologicalOrder() {
  const visited = new Set();
  const order = [];

  function visit(id) {
    if (visited.has(id)) return; 
    visited.add(id);
    const svc = SERVICE_DEFINITIONS[id];
    if (svc) {
      for (const dep of svc.dependencies) {
        visit(dep);
      }
    }
    order.push(id);
  }

  for (const id of Object.keys(SERVICE_DEFINITIONS)) {
    visit(id);
  }
  return order;
}


function dependencyDepth(serviceId, _visited) {
  const seen = _visited || new Set();
  if (seen.has(serviceId)) return 0;
  seen.add(serviceId);
  const svc = SERVICE_DEFINITIONS[serviceId];
  if (!svc) return -1;
  if (svc.dependencies.length === 0) return 0;
  let maxDepth = 0;
  for (const dep of svc.dependencies) {
    const d = dependencyDepth(dep, seen);
    if (d > maxDepth) maxDepth = d;
  }
  return maxDepth + 1;
}

function serviceCount() {
  return Object.keys(SERVICE_DEFINITIONS).length;
}

// Backwards-compatible flat export — expanded to include all 8 services
module.exports = {
  gateway: { id: 'gateway', port: 8090 },
  routing: { id: 'routing', port: 8091 },
  policy: { id: 'policy', port: 8092 },
  resilience: { id: 'resilience', port: 8093 },
  analytics: { id: 'analytics', port: 8094 },
  audit: { id: 'audit', port: 8095 },
  notifications: { id: 'notifications', port: 8096 },
  security: { id: 'security', port: 8097 },
  SERVICE_DEFINITIONS,
  REQUIRED_COMMAND_FIELDS,
  getServiceUrl,
  validateContract,
  topologicalOrder,
  dependencyDepth,
  serviceCount,
};

File: go/ironfleet/shared/contracts/contracts.go
package contracts

import "fmt"

// ---------------------------------------------------------------------------
// Service definitions
// ---------------------------------------------------------------------------

type ServiceDefinition struct {
	ID           string
	Port         int
	HealthPath   string
	Version      string
	Dependencies []string
}

var ServiceDefs = map[string]ServiceDefinition{
	"gateway":       {ID: "gateway", Port: 8130, HealthPath: "/health", Version: "1.0.0", Dependencies: []string{"routing", "policy"}},
	"routing":       {ID: "routing", Port: 8131, HealthPath: "/health", Version: "1.0.0", Dependencies: []string{"policy"}},
	"policy":        {ID: "policy", Port: 8132, HealthPath: "/health", Version: "1.0.0"},
	"resilience":    {ID: "resilience", Port: 8133, HealthPath: "/health", Version: "1.0.0", Dependencies: []string{"policy"}},
	"analytics":     {ID: "analytics", Port: 8134, HealthPath: "/health", Version: "1.0.0", Dependencies: []string{"routing"}},
	"audit":         {ID: "audit", Port: 8135, HealthPath: "/health", Version: "1.0.0"},
	"notifications": {ID: "notifications", Port: 8136, HealthPath: "/health", Version: "1.0.0", Dependencies: []string{"policy"}},
	"security":      {ID: "security", Port: 8137, HealthPath: "/health", Version: "1.0.0"},
}

// Backwards-compatible flat map — includes all 8 services
var Contracts = map[string]map[string]interface{}{
	"gateway":       {"id": "gateway", "port": 8130},
	"routing":       {"id": "routing", "port": 8131},
	"policy":        {"id": "policy", "port": 8132},
	"resilience":    {"id": "resilience", "port": 8133},
	"analytics":     {"id": "analytics", "port": 8134},
	"audit":         {"id": "audit", "port": 8135},
	"notifications": {"id": "notifications", "port": 8136},
	"security":      {"id": "security", "port": 8137},
}


var RequiredCommandFields = []string{"ID", "Action", "Service"}

// ---------------------------------------------------------------------------
// Service URL resolution
// ---------------------------------------------------------------------------


func GetServiceURL(serviceID, baseDomain string) string {
	svc, ok := ServiceDefs[serviceID]
	if !ok {
		return ""
	}
	if baseDomain == "" {
		baseDomain = "localhost"
	}
	return fmt.Sprintf("http://%s:%d%s", baseDomain, svc.Port, svc.HealthPath)
}

// ---------------------------------------------------------------------------
// Service count
// ---------------------------------------------------------------------------

func ServiceCount() int {
	return len(ServiceDefs)
}

// ---------------------------------------------------------------------------
// Dependency depth
// ---------------------------------------------------------------------------


func DependencyDepth(serviceID string) int {
	svc, ok := ServiceDefs[serviceID]
	if !ok {
		return 0
	}
	if len(svc.Dependencies) == 0 {
		return 0
	}
	maxDepth := 0
	for _, dep := range svc.Dependencies {
		d := DependencyDepth(dep)
		if d > maxDepth {
			maxDepth = d
		}
	}
	return maxDepth + 1
}

// ---------------------------------------------------------------------------
// Contract validation
// ---------------------------------------------------------------------------

type ValidationResult struct {
	Valid   bool
	Reason  string
	Service *ServiceDefinition
}


func ValidateContract(serviceID string) ValidationResult {
	svc, ok := ServiceDefs[serviceID]
	if !ok {
		return ValidationResult{Valid: false, Reason: "unknown_service"}
	}
	if svc.Port <= 1024 {
		return ValidationResult{Valid: false, Reason: "invalid_port"}
	}
	return ValidationResult{Valid: true, Service: &svc}
}

// ---------------------------------------------------------------------------
// Topological ordering
// ---------------------------------------------------------------------------


func TopologicalOrder() []string {
	visited := make(map[string]bool)
	order := make([]string, 0)

	var visit func(id string)
	visit = func(id string) {
		if visited[id] {
			return
		}
		visited[id] = true
		if svc, ok := ServiceDefs[id]; ok {
			for _, dep := range svc.Dependencies {
				visit(dep)
			}
		}
		order = append(order, id)
	}

	for id := range ServiceDefs {
		visit(id)
	}
	return order
}
